keps/sig-autoscaling/draft-20180419-enhance-hpa-metrics-specificity.md

```diff
@@ -0,0 +1,247 @@
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Drawbacks [optional]](#drawbacks-optional)
+* [Alternatives [optional]](#alternatives-optional)
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+
+The `External` metric source type in the HPA currently supports passing
+a metric label selectors, which is passed to the custom metrics API
+(custom.metrics.k8s.io) to select a more specific metrics series.  This
+allows users to more easily make use of existing metrics structure,
+without need to manipulate their metrics labeling and ingestion
+externally.
+
+Additionally, it supports the `targetAverageValue` field, which allows
+artificially dividing an external metric by the number of replicas in the
+target scalable.
+
+This proposal brings both of those fields to the `Object` metric source
+type, and further brings the selector field to the `Pods` metric source
+type, making both types more flexible and bringing them in line with the
+`External` metrics types.
+
+## Motivation
+
+With custom-metrics-based autoscaling, users frequently ask how to select
+more specific metrics in their metric storage.  For instance, a user might
+have message queue statefulset with several queues producing metrics:
+
+```
+queue_length{statefulset="foo",pod="foo-1",queue="some-jobs"}
+queue_length{statefulset="foo",pod="foo-1",queue="other-jobs"}
+```
+
+Suppose they have a pool of works that they wish to scale for each queue.
+In the current-day HPA, it's non-trivial to allow selecting the metric for
+a specific queue.  Current suggestions are metric-backend-specific (for
+instance, you could create a Prometheus recording rule to relabel or
+rename the metric), and often involve making external changes to the
+metrics pipeline.
+
+With the addition of the metrics label selector, users could simply select
+the queue using the label selector:
+
+```yaml
+- type: Object
+  object:
+    target:
+      kind: StatefulSet
+      apiVersion: apps/v1
+      name: foo
+    targetValue: 2
+    metric:
+      name: queue_length
+      selector: "queue=some-jobs"
+```
+
+Similarly, in discussions of scaling on queues, being able to divide
+a target backlog length by the number of available pods is often useful --
+for instance, a backlog length of 3 might be acceptable if there are three
+pods processing items, but not if there is only one.
+
+### Goals
+
+- The autoscaling/v2 API is updated with the additional fields
+  described below.
+- A corresponding change is made to the custom metrics API to support the
+  additional label selector.
+- The testing adapter is updated to support these changes (for e2e
+  purposes).
+
+### Non-Goals
+
+It is outside of the purview of the KEP to ensure that current custom
+metrics adapters support the new changes -- this is up to those adapters
+maintainers.
+
+## Proposal
```

Comment:
I think this misses a single part - what do we do if multiple time series are matched by selector? In external metrics API we sum all the values (we assume the user may not want to drill down on some dimension captured by metric, for example use all HTTP requests, regardless of response code). My assumption is we want to do the same? We should say it explicitly though.

Reply:
If multiple series are matched, the adapter must deal with it, similarly to the current CM API (see my other comment).